invoke-ai/InvokeAI
# ui: multi-select and batched gallery image operations

## What type of PR is this? (check all applicable)



## Have you discussed this change with the InvokeAI team?

      
## Have you updated all relevant documentation?


## Description

[feat: add multi-select to gallery](https://github.com/invoke-ai/InvokeAI/pull/4032/commits/c3e99ad145e73a5d836300c70bd81aeb5b574a7c)
[c3e99ad](https://github.com/invoke-ai/InvokeAI/pull/4032/commits/c3e99ad145e73a5d836300c70bd81aeb5b574a7c)

multi-select actions include:
- drag to board to move all to that board
- right click to add all to board or delete all

backend changes:
- add routes for changing board for list of image names, deleting list of images
- change image-specific routes to `images/i/{image_name}` to not clobber other routes (like `images/upload`, `images/delete`)
- subclass pydantic `BaseModel` as `BaseModelExcludeNull`, which excludes null values when calling `dict()` on the model. this fixes inconsistent types related to JSON parsing null values into `null` instead of `undefined`
- remove `board_id` from `remove_image_from_board`

frontend changes:
- multi-selection stuff uses `ImageDTO[]` as payloads, for dnd and other mutations. this gives us access to image `board_id`s when hitting routes, and enables efficient cache updates.
- consolidate change board and delete image modals to handle single and multiples
- board totals are now re-fetched on mutation and not kept in sync manually - was way too tedious to do this
- fixed warning about nested `<p>` elements
- closes #4088 , need to handle case when `autoAddBoardId` is `"none"`
- add option to show gallery image delete button on every gallery image

frontend refactors/organisation:
- make typegen script js instead of ts
- enable `noUncheckedIndexedAccess` to help avoid bugs when indexing into arrays, many small changes needed to satisfy TS after this
- move all image-related endpoints into `endpoints/images.ts`, its a big file now, but this fixes a number of circular dependency issues that were otherwise felt impossible to resolve

## Related Tickets & Documents

- Supersedes #3669 #3771
- Closes #3882 #4088

## QA Instructions, Screenshots, Recordings

Basic functionality to review:
- Use shift and ctrl/cmd to select multiple images
- Drag and drop images between boards
- Right click multi-selection to change the board for a group of images
- Right click multi-selection to delete many images at once
- General drive-through of gallery features & auto-add

Specific cases:
- Deleting a single image should still work as before
- Deleting a single image that is in use in another feature should prompt you and reset that feature when deletion is confirmed
- Same as above for multiple images
- Adding a single image to a board should still work as before
- Adding multiple images to a board should work via context menu and dnd

There's a lot to test here and some complicated cache updates, so it's hard to test all the possible sequences of events.



## Added/updated tests?


## [optional] Are there any post deployment tasks we need to perform?
